contracts/ERC721/internal/properties/ERC721BasicProperties.sol

```diff
@@ -0,0 +1,131 @@
+pragma solidity ^0.8.13;
+
+import "../util/ERC721TestBase.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+abstract contract CryticERC721BasicProperties is CryticERC721TestBase {
+    using Address for address;
+
+    ////////////////////////////////////////
+    // Properties
+
+    // Querying the balance of address(0) should throw
+    function test_ERC20_balanceOfZeroAddressMustRevert() public virtual {
+        balanceOf(address(0));
+        assertWithMsg(false, "address(0) balance query should have reverted");
+    }
+
+    // Querying the owner of an invalid token should throw
+    function test_ERC721_ownerOfInvalidTokenMustRevert(uint256 tokenId) public virtual {
+        require(!_exists(tokenId));
+        ownerOf(tokenId);
+        assertWithMsg(false, "Invalid token owner query should have reverted");
+    }
+
+    // Approving an invalid token should throw
+    function test_ERC721_approvingInvalidTokenMustRevert(uint256 tokenId) public virtual {
+        require(!_exists(tokenId));
+        approve(address(0), tokenId);
+        assertWithMsg(false, "Approving an invalid token should have reverted");
+    }
+
+    // transferFrom a token that the caller is not approved for should revert
+    function test_ERC721_transferFromNotApproved(address target) public virtual {
+        uint256 selfBalance = balanceOf(target);
+        require(selfBalance > 0);        
+        require(target != address(this));
+        require(target != msg.sender);
+        uint tokenId = tokenOfOwnerByIndex(target, 0);
+        _approve(address(0), tokenId);
+        _setApprovalForAll(target, msg.sender, false);
+        require(ownerOf(tokenId) == target);
+
+        transferFrom(target, msg.sender, tokenId);
+        assertWithMsg(ownerOf(tokenId) == target, "Target");
+        assertWithMsg(ownerOf(tokenId) == msg.sender, "Transferred a token without being approved.");
+        
+    }
+
+    // transferFrom should reset approval for that token
+    function test_ERC721_transferFromResetApproval(address target) public virtual {
+        uint256 selfBalance = balanceOf(msg.sender);
+        require(selfBalance > 0);  
+        require(target != address(this));
+        require(target != msg.sender);
+        uint tokenId = tokenOfOwnerByIndex(msg.sender, 0);
+        transferFrom(msg.sender, target, tokenId);
+
+        address approved = getApproved(tokenId);
+        assertWithMsg(approved == address(0), "Approval was not reset");
+    }
+
+    // transferFrom correctly updates owner
+    function test_ERC721_transferFromUpdatesOwner(address target) public virtual {
+        uint256 selfBalance = balanceOf(msg.sender);
+        require(selfBalance > 0);  
+        require(target != address(this));
+        require(target != msg.sender);
+        uint tokenId = tokenOfOwnerByIndex(msg.sender, 0);
+        transferFrom(msg.sender, target, tokenId);
```

Comment:
Use external call (and prank) with try/catch and `assert(false)` in case of revert.